core/lib/multivm/src/versions/vm_fast/vm.rs

```diff
@@ -744,20 +782,23 @@ impl<S: ReadStorage, T: Tracer> World<S, T> {
 }
 
 impl<S: ReadStorage, T: Tracer> zksync_vm2::StorageInterface for World<S, T> {
-    fn read_storage(&mut self, contract: H160, key: U256) -> Option<U256> {
+    fn read_storage(&mut self, contract: H160, key: U256) -> StorageSlot {
         let key = &StorageKey::new(AccountTreeId::new(contract), u256_to_h256(key));
-        if self.storage.is_write_initial(key) {
-            None
-        } else {
-            Some(self.storage.read_value(key).as_bytes().into())
+        let value = U256::from_big_endian(self.storage.read_value(key).as_bytes());
+        let is_write_initial = self.storage.is_write_initial(key);
+        StorageSlot {
+            value,
+            is_write_initial,
         }
     }
 
-    fn cost_of_writing_storage(&mut self, initial_value: Option<U256>, new_value: U256) -> u32 {
-        let is_initial = initial_value.is_none();
-        let initial_value = initial_value.unwrap_or_default();
+    fn read_storage_value(&mut self, contract: H160, key: U256) -> U256 {
+        let key = &StorageKey::new(AccountTreeId::new(contract), u256_to_h256(key));
+        U256::from_big_endian(self.storage.read_value(key).as_bytes())
+    }
 
-        if initial_value == new_value {
+    fn cost_of_writing_storage(&mut self, initial_slot: StorageSlot, new_value: U256) -> u32 {
```

Comment:
Not sure if bug or poor naming. If initial_slot contains the initial value of the slot it doesn't make sense that it is a StorageSlot because it will always be initial according to the name.

Reply:
Bad naming; "initial" here means that the provided value is the initial value. I'll shorten the name to `slot`.